providers/apis/defi/ethmulticlient/multi_client.go

```diff
@@ -87,77 +89,103 @@ func NewMultiRPCClientFromEndpoints(
 
 // BatchCallContext injects a call to eth_blockNumber, and makes batch calls to the underlying EVMClients.
 // It returns the first response it sees from a node which has the greatest height. An error is returned
-// only when all clients fail.
+// only when no client was able to successfully provide a height or errored when sending the BatchCall.
 func (m *MultiRPCClient) BatchCallContext(ctx context.Context, batchElems []rpc.BatchElem) error {
 	if len(batchElems) == 0 {
 		m.logger.Debug("BatchCallContext called with 0 elems")
 		return nil
 	}
 
-	req := make([]rpc.BatchElem, len(batchElems)+1)
-	copy(req, batchElems)
+	// define a result struct that go routines will populate and append to a slice when they complete their request.
+	type result struct {
+		height  uint64
+		results []rpc.BatchElem
+	}
+	results := make([]result, len(m.clients))
+
+	// error slice to capture errors go routines encounter.
+	errs := make([]error, len(m.clients))
+	wg := new(sync.WaitGroup)
+	// this is the index of where we will have an eth_blockNumber call.
 	blockNumReqIndex := len(batchElems)
-	req[blockNumReqIndex] = EthBlockNumberBatchElem()
-	errs := fmt.Errorf("all eth client requests failed")
+	// for each client, spin up a go routine that executes a BatchCall.
+	for clientIdx, client := range m.clients {
+		wg.Add(1)
+		go func(i int) {
+			url := m.api.Endpoints[i].URL
+
+			// append an eth_blockNumber call to the requests. we do this because we want the greatest height results only.
+			req := make([]rpc.BatchElem, len(batchElems)+1)
+			copy(req, batchElems)
+			req[blockNumReqIndex] = EthBlockNumberBatchElem()
+
+			err := client.BatchCallContext(ctx, req)
+
+			// if there was an error, or if the block_num request didn't have result / errored
+			// we log the error and append to error slice.
+			if err != nil || req[blockNumReqIndex].Result == "" || req[blockNumReqIndex].Error != nil {
+				errs[i] = fmt.Errorf("endpoint request failed: %w, %w", err, req[blockNumReqIndex].Error)
+				m.logger.Debug(
+					"endpoint request failed",
+					zap.Error(err),
+					zap.Any("result", req[blockNumReqIndex].Result),
+					zap.Error(req[blockNumReqIndex].Error),
+					zap.String("url", url),
+				)
+			} else { // the batch call succeeded, and the eth_blockNumber call had results.\
+				// try to get the block number.
+				r, ok := req[blockNumReqIndex].Result.(*string)
+				if !ok {
+					errs[i] = fmt.Errorf("result from eth_blockNumber was not a string")
+					m.logger.Debug(
+						"result from eth_blockNumber was not a string",
+						zap.String("url", url),
+					)
+				} else {
```

Comment:
this is kind of a nasty if/else chain.. which could be solved by using errgroup, however errgroup only gives you the first non-nil error, instead of all of them, which the code before did. so opting to keep that the same here.

but if having _all_ the errors is not that important, i can clean this up by using errgroup. that i way i can just log and then return err when needed, reducing a LOT of if/else junk.